reasoner/planner/AnswerCountEstimator.java

```diff
@@ -60,210 +59,195 @@
+            Variable bestScalingVar = scale.second().orElse(null);
+            // Update the index
+            model.variables.forEach(v -> modelsWithVar.computeIfAbsent(v, v1 -> new HashSet<>()).add(model));
+
+            // Find scaling factor
+            for (Variable v : model.variables) {
+                double ans = (double) model.estimateAnswers(set(v));
+                if (bestUnary.containsKey(v)) {
+                    if (ans < bestUnary.get(v)) {
+                        unaryUpdates.put(v, ans);
+                    } else if (bestUnary.get(v) / ans < bestScaler) {
+                        bestScaler = bestUnary.get(v) / ans;
+                        bestScalingVar = v;
+                    }
+                } else {
+                    unaryUpdates.put(v, ans);
+                }
+            }
 
-        private Concludable getConcludable(ResolvableConjunction conjunction) {
-            return concludables.get(conjunction);
+            // Find cascading effects of scaling on other vars (to propagate)
+            if (bestScalingVar != null) {
+                modelScale.put(model, new Pair<>(bestScaler, Optional.of(bestScalingVar)));
+                for (Variable v : model.variables) {
+                    if (v == bestScalingVar) continue;
+                    double newUnary = scaledEstimate(model, modelScale.get(model), set(v));
+                    if (newUnary < bestUnary.getOrDefault(v, Double.MAX_VALUE)) {
+                        unaryUpdates.put(v, Math.min(newUnary, unaryUpdates.getOrDefault(v, Double.MAX_VALUE)));
+                    }
+                }
+            } else {
+                assert bestScaler == 1;
+                modelScale.put(model, new Pair<>(bestScaler, Optional.empty()));
+            }
         }
 
-        public List<ResolvableConjunction> peekUntil(ResolvableConjunction untilInclusive) {
-            assert stack.contains(untilInclusive);
-            List<ResolvableConjunction> subList = new ArrayList<>();
-            for (int i = stack.size() - 1; stack.get(i) != untilInclusive; i--) {
-                subList.add(stack.get(i));
+        private void propagate(Map<Variable, Double> unaryUpdates) {
+            // Ideally, we'd just remove and add each model again till unaryUpdates is empty.
+            int maxIters = 2 * modelScale.size(); // TODO: Consider pruning out small changes
+            while (!unaryUpdates.isEmpty() && maxIters > 0) {
+                Set<LocalModel> affectedModels = iterate(unaryUpdates.keySet()).flatMap(v -> iterate(modelsWithVar.get(v))).toSet();
+                assert iterate(unaryUpdates.entrySet()).allMatch(update -> update.getValue() < bestUnary.getOrDefault(update.getKey(), Double.MAX_VALUE));
+                bestUnary.putAll(unaryUpdates);
+                unaryUpdates.clear();
+                affectedModels.forEach(model -> addAndCollectUnaryUpdates(model, unaryUpdates));
+                maxIters--;
             }
```

Comment:
When adding a model causes a revision in the best-known estimate, we updated the estimate and 'propagate' the change to other models by recomputing their scaling-factor based on the new estimate. This is the same process as adding them for the first time.
Since one bound reducing can lead to other ones being reduced and so on, we restrict the number of iterations of the propagation.